File: packages/health_connector_core/lib/src/models/exceptions/health_connector_exception.dart
import 'package:health_connector_core/src/annotations/annotations.dart'
    show sinceV1_0_0, sinceV2_0_0, sinceV3_0_0;
import 'package:health_connector_core/src/models/exceptions/health_connector_error_code.dart'
    show HealthConnectorErrorCode;
import 'package:meta/meta.dart' show immutable;

/// Sealed exception class for all health connector plugin errors.
///
/// {@category Exceptions}
@sinceV1_0_0
@immutable
sealed class HealthConnectorException implements Exception {
  /// Creates a [HealthConnectorException].
  ///
  /// ## Parameters
  /// - [code]: The error code identifying the type of error.
  /// - [message]: A human-readable description of the error.
  /// - [cause]: The underlying cause of this exception, if any.
  const HealthConnectorException(
    this.code,
    this.message, {
    this.cause,
    this.stackTrace,
  });

  /// Creates a specific [HealthConnectorException] subclass based on the
  /// provided [code].
  ///
  /// ## Parameters
  /// - [code]: The error code identifying the type of error.
  /// - [message]: A human-readable description of the error.
  /// - [cause]: The underlying cause of this exception, if any.
  /// - [stackTrace]: The stack trace of this exception, if any.
  @sinceV2_0_0
  factory HealthConnectorException.fromCode(
    HealthConnectorErrorCode code,
    String message, {
    Object? cause,
    StackTrace? stackTrace,
  }) {
    switch (code) {
      // Authorization exceptions
      case HealthConnectorErrorCode.permissionNotGranted:
        return AuthorizationException(
          code,
          message,
          cause: cause,
          stackTrace: stackTrace,
        );

      // Configuration exception
      case HealthConnectorErrorCode.permissionNotDeclared:
        return ConfigurationException(
          message,
          cause: cause,
          stackTrace: stackTrace,
        );

      // Invalid argument exception
      case HealthConnectorErrorCode.invalidArgument:
        return InvalidArgumentException(
          message,
          cause: cause,
          stackTrace: stackTrace,
        );

      // Health service unavailable exceptions
      case HealthConnectorErrorCode.healthServiceUnavailable:
      case HealthConnectorErrorCode.healthServiceRestricted:
      case HealthConnectorErrorCode.healthServiceNotInstalledOrUpdateRequired:
        return HealthServiceUnavailableException(
          code,
          message,
          cause: cause,
          stackTrace: stackTrace,
        );

      // Health service exceptions
      case HealthConnectorErrorCode.healthServiceDatabaseInaccessible:
      case HealthConnectorErrorCode.ioError:
      case HealthConnectorErrorCode.remoteError:
      case HealthConnectorErrorCode.rateLimitExceeded:
      case HealthConnectorErrorCode.dataSyncInProgress:
        return HealthServiceException(
          code,
          message,
          cause: cause,
          stackTrace: stackTrace,
        );

      // Unsupported operation exception
      case HealthConnectorErrorCode.unsupportedOperation:
        return UnsupportedOperationException(
          message,
          cause: cause,
          stackTrace: stackTrace,
        );

      // Unknown exception
      case HealthConnectorErrorCode.unknownError:
        return UnknownException(
          message,
          cause: cause,
          stackTrace: stackTrace,
        );
    }
  }

  /// The error code identifying the type of error.
  final HealthConnectorErrorCode code;

  /// A human-readable description of the error.
  final String message;

  /// The underlying cause of this exception, if any.
  final Object? cause;

  /// The stack trace of this exception, if any.
  final StackTrace? stackTrace;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is HealthConnectorException &&
          runtimeType == other.runtimeType &&
          code == other.code &&
          message == other.message &&
          cause == other.cause &&
          stackTrace == other.stackTrace;

  @override
  int get hashCode =>
      code.hashCode ^
      message.hashCode ^
      (cause?.hashCode ?? 0) ^
      (stackTrace?.hashCode ?? 0);

  @override
  String toString() {
    final buffer = StringBuffer(
      'HealthConnectorException(code=${code.code}, message=$message',
    );

    if (cause != null) {
      buffer.write(', cause=$cause');
    }

    if (stackTrace != null) {
      buffer.write(', stackTrace=$stackTrace');
    }

    buffer.write(')');

    return buffer.toString();
  }
}

/// Exception thrown when authorization/permission issues occur.
///
/// This exception groups all authorization-related errors:
/// - [HealthConnectorErrorCode.permissionNotGranted]

///
/// {@category Exceptions}
@sinceV3_0_0
@immutable
final class AuthorizationException extends HealthConnectorException {
  /// Creates an [AuthorizationException].
  const AuthorizationException(
    super.code,
    super.message, {
    super.cause,
    super.stackTrace,
  });
}

/// Exception thrown when app configuration is missing or invalid.
///
/// This exception is for configuration errors:
/// - [HealthConnectorErrorCode.permissionNotDeclared]
///
/// {@category Exceptions}
@sinceV3_0_0
@immutable
final class ConfigurationException extends HealthConnectorException {
  /// Creates a [ConfigurationException].
  const ConfigurationException(
    String message, {
    Object? cause,
    StackTrace? stackTrace,
  }) : super(
         HealthConnectorErrorCode.permissionNotDeclared,
         message,
         cause: cause,
         stackTrace: stackTrace,
       );
}

/// Exception thrown when an invalid argument is provided to the API.
///
/// ## See Also
///
/// - [HealthConnectorErrorCode.invalidArgument]
///
/// {@category Exceptions}
@sinceV2_0_0
@immutable
final class InvalidArgumentException extends HealthConnectorException {
  /// Creates an [InvalidArgumentException].
  const InvalidArgumentException(
    String message, {
    Object? cause,
    StackTrace? stackTrace,
  }) : super(
         HealthConnectorErrorCode.invalidArgument,
         message,
         cause: cause,
         stackTrace: stackTrace,
       );
}

/// Exception thrown when health service is unavailable.
///
/// This exception groups service unavailability errors:
/// - [HealthConnectorErrorCode.healthServiceUnavailable]
/// - [HealthConnectorErrorCode.healthServiceRestricted]
/// - [HealthConnectorErrorCode.healthServiceNotInstalledOrUpdateRequired]
///
/// {@category Exceptions}
@sinceV3_0_0
@immutable
final class HealthServiceUnavailableException extends HealthConnectorException {
  /// Creates a [HealthServiceUnavailableException].
  const HealthServiceUnavailableException(
    super.code,
    super.message, {
    super.cause,
    super.stackTrace,
  });
}

/// Exception thrown when health service operational errors occur.
///
/// This exception groups operational errors:
/// - [HealthConnectorErrorCode.healthServiceDatabaseInaccessible]
/// - [HealthConnectorErrorCode.ioError]
/// - [HealthConnectorErrorCode.remoteError]
/// - [HealthConnectorErrorCode.rateLimitExceeded]
/// - [HealthConnectorErrorCode.dataSyncInProgress]
///
/// {@category Exceptions}
@sinceV3_0_0
@immutable
final class HealthServiceException extends HealthConnectorException {
  /// Creates a [HealthServiceException].
  const HealthServiceException(
    super.code,
    super.message, {
    super.cause,
    super.stackTrace,
  });
}

/// Exception thrown when an API not supported by the current platform or
/// version is used.
///
/// ## See Also
///
/// - [HealthConnectorErrorCode.unsupportedOperation]
///
/// {@category Exceptions}
@sinceV2_0_0
@immutable
final class UnsupportedOperationException extends HealthConnectorException {
  /// Creates an [UnsupportedOperationException].
  const UnsupportedOperationException(
    String message, {
    Object? cause,
    StackTrace? stackTrace,
  }) : super(
         HealthConnectorErrorCode.unsupportedOperation,
         message,
         cause: cause,
         stackTrace: stackTrace,
       );
}

/// Exception thrown when an unknown or unexpected error occurred.
///
/// ## See Also
///
/// - [HealthConnectorErrorCode.unknownError]
///
/// {@category Exceptions}
@sinceV2_0_0
@immutable
final class UnknownException extends HealthConnectorException {
  /// Creates an [UnknownException].
  const UnknownException(
    String message, {
    Object? cause,
    StackTrace? stackTrace,
  }) : super(
         HealthConnectorErrorCode.unknownError,
         message,
         cause: cause,
         stackTrace: stackTrace,
       );
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/exceptions/HealthConnectorException.kt
package com.phamtunglam.health_connector_hc_android.exceptions

import com.phamtunglam.health_connector_hc_android.pigeon.HealthConnectorErrorCodeDto

/**
 * Defines a standardized set of errors that can occur.
 *
 * This sealed class hierarchy provides type-safe error handling with associated context
 * and debugging information. Each error type contains relevant details for proper error
 * handling and debugging.
 */
sealed class HealthConnectorException : Throwable() {
    /**
     * Error code.
     */
    abstract val code: HealthConnectorErrorCodeDto

    /**
     * Optional additional key-value data for debugging purposes.
     */
    abstract val context: Map<String, Any>?

    /**
     * Authorization-related errors.
     *
     * Covers permission denial, not-determined state, and guest mode restrictions.
     *
     * Error codes:
     * - PERMISSION_NOT_GRANTED
     * - AUTHORIZATION_NOT_DETERMINED
     * - GUEST_MODE_NOT_PERMITTED
     *
     * @property code Specific authorization error code.
     * @property message Description of the authorization error.
     * @property cause Optional underlying exception.
     * @property context Additional debugging information.
     */
    data class Authorization(
        override val code: HealthConnectorErrorCodeDto,
        override val message: String,
        override val cause: Throwable? = null,
        override val context: Map<String, Any>? = null,
    ) : HealthConnectorException()

    /**
     * Configuration-related errors.
     *
     * Indicates missing permissions in AndroidManifest.xml or other configuration issues.
     *
     * Error codes:
     * - PERMISSION_NOT_DECLARED
     *
     * @property message Description of the configuration issue.
     * @property cause Optional underlying exception.
     * @property context Additional debugging information about the missing configuration.
     */
    data class Configuration(
        override val message: String,
        override val cause: Throwable? = null,
        override val context: Map<String, Any>? = null,
    ) : HealthConnectorException() {
        override val code: HealthConnectorErrorCodeDto =
            HealthConnectorErrorCodeDto.PERMISSION_NOT_DECLARED
    }

    /**
     * Health service unavailability errors.
     *
     * Covers cases where the health service cannot be accessed.
     *
     * Error codes:
     * - HEALTH_SERVICE_UNAVAILABLE
     * - HEALTH_SERVICE_RESTRICTED
     * - HEALTH_SERVICE_NOT_INSTALLED_OR_UPDATE_REQUIRED
     *
     * @property code Specific service unavailability error code.
     * @property message Description of why the health service is unavailable.
     * @property cause Optional underlying exception.
     * @property context Additional debugging information.
     */
    data class HealthServiceUnavailable(
        override val code: HealthConnectorErrorCodeDto,
        override val message: String,
        override val cause: Throwable? = null,
        override val context: Map<String, Any>? = null,
    ) : HealthConnectorException()

    /**
     * Health service operational errors.
     *
     * Covers transient errors during health service operations.
     *
     * Error codes:
     * - HEALTH_SERVICE_DATABASE_INACCESSIBLE
     * - IO_ERROR
     * - REMOTE_ERROR
     * - RATE_LIMIT_EXCEEDED
     * - DATA_SYNC_IN_PROGRESS
     *
     * @property code Specific service error code.
     * @property message Description of the service error.
     * @property cause Optional underlying exception.
     * @property context Additional debugging information.
     */
    data class HealthService(
        override val code: HealthConnectorErrorCodeDto,
        override val message: String,
        override val cause: Throwable? = null,
        override val context: Map<String, Any>? = null,
    ) : HealthConnectorException()

    /**
     * Invalid argument errors.
     *
     * Signals that a method was called with an invalid argument.
     *
     * This can occur when:
     * - startTime is after endTime
     * - Value is out of valid range
     * - Record ID does not exist
     * - Malformed record data
     * - Expired change token
     *
     * Error codes:
     * - INVALID_ARGUMENT
     *
     * @property message Description of the invalid argument.
     * @property cause Optional underlying exception.
     * @property context Additional debugging information about the invalid argument.
     */
    data class InvalidArgument(
        override val message: String,
        override val cause: Throwable? = null,
        override val context: Map<String, Any>? = null,
    ) : HealthConnectorException() {
        override val code: HealthConnectorErrorCodeDto =
            HealthConnectorErrorCodeDto.INVALID_ARGUMENT
    }

    /**
     * Unsupported operation errors.
     *
     * Indicates that the requested operation is not supported.
     *
     * This can occur when:
     * - Calling a platform-specific API on the wrong platform
     * - Requesting a data type unsupported by the current SDK version
     *
     * Error codes:
     * - UNSUPPORTED_OPERATION
     *
     * @property message Description of the unsupported operation.
     * @property cause Optional underlying exception.
     * @property context Additional debugging information.
     */
    data class UnsupportedOperation(
        override val message: String,
        override val cause: Throwable? = null,
        override val context: Map<String, Any>? = null,
    ) : HealthConnectorException() {
        override val code: HealthConnectorErrorCodeDto =
            HealthConnectorErrorCodeDto.UNSUPPORTED_OPERATION
    }

    /**
     * Unknown errors.
     *
     * A generic, unexpected error that doesn't fit other categories.
     *
     * This should be used as a fallback for unmapped error codes or unexpected situations.
     *
     * Error codes:
     * - UNKNOWN_ERROR
     *
     * @property message Description of the unknown error.
     * @property cause Optional underlying exception.
     * @property context Additional debugging information.
     */
    data class Unknown(
        override val message: String,
        override val cause: Throwable? = null,
        override val context: Map<String, Any>? = null,
    ) : HealthConnectorException() {
        override val code: HealthConnectorErrorCodeDto = HealthConnectorErrorCodeDto.UNKNOWN_ERROR
    }
}
